### PR TITLE
BaselineNeonErrorFormatter: Sort errors by normalized relative path, …

### PR DESCRIPTION
**Pull request by dktapps**
…instead of original absolute path

while in theory the net result should be the same (because the leading part of the path would be the same), / and \ have a different sorting order, so output gets sorted differently on Windows than on Linux.
This fixes phpstan/phpstan#5085.

### CODE REVIEW

**Comment by mvorisek** [comment]
LGTM

**Comment by mvorisek** [comment]
LGTM, adding `SORT_NATURAL` flag to the sort functions would be also helful so mixed case is sorted natuarally

**Comment by ondrejmirtes** [comment]
Any change in src/ without change in tests/ is a red flag. Add a failing test first, thanks.

**Comment by ondrejmirtes** [comment]
@dktapps Do you plan to finish this PR? Thanks.

**Comment by dktapps** [comment]
Sorry, I forgot all about it. I'll try and find some time to finish it today.

**Comment by staabm** [comment]
> Add a failing test first

@dktapps I think I have a failling test for the issue, but the fix proposed here seems to not fix it https://github.com/phpstan/phpstan-src/pull/667

**Comment by dktapps** [comment]
As I recall, last time I tried to write tests with this, I had some problem with SimpleRelativePathHelper not normalizing the separator.

File: src/Command/ErrorFormatter/BaselineNeonErrorFormatter.php [hunk 1]
```diff
@@ -40,7 +40,7 @@ public function formatErrors(
 			if (!$fileSpecificError->canBeIgnored()) {
 				continue;
 			}
-			$fileErrors[$fileSpecificError->getFilePath()][] = $fileSpecificError->getMessage();
+			$fileErrors[$this->relativePathHelper->getRelativePath($fileSpecificError->getFilePath())][] = $fileSpecificError->getMessage();
```
Comment (mvorisek) [inline src/Command/ErrorFormatter/BaselineNeonErrorFormatter.php:43]:
```suggestion
			$fileErrors[str_replace('\\', '/', $this->relativePathHelper->getRelativePath($fileSpecificError->getFilePath()))][] = $fileSpecificError->getMessage();
```

**Comment by dktapps** [comment]
I've added a test case. If you run it against the original code, the `Error without path separators` will appear in between the other two, like so: 
image

With the change it passes as expected.

**Review by mvorisek — APPROVED** [review]
LGTM

File: tests/PHPStan/Command/ErrorFormatter/BaselineNeonErrorFormatterTest.php [hunk 2]
```diff
@@ -207,6 +207,9 @@ public function outputOrderingProvider(): \Generator
 			new Error('Second error in a different file', 'TestfileB', 4),
 			new Error('Error #1 in a different file', 'TestfileB', 5),
 			new Error('Second error in a different file', 'TestfileB', 6),
+			new Error('Error with Windows path separators', 'TestFiles\\TestA', 1),
```
Comment (mvorisek) [inline tests/PHPStan/Command/ErrorFormatter/BaselineNeonErrorFormatterTest.php:210]: It should be named a "dir separator", not a "path separator" as the later means something else, see https://www.php.net/manual/en/dir.constants.php

File: src/File/SimpleRelativePathHelper.php [hunk 3]
```diff
@@ -15,10 +15,10 @@ public function __construct(string $currentWorkingDirectory)
 	public function getRelativePath(string $filename): string
 	{
 		if ($this->currentWorkingDirectory !== '' && strpos($filename, $this->currentWorkingDirectory) === 0) {
-			return substr($filename, strlen($this->currentWorkingDirectory) + 1);
+			return str_replace('\\', '/', substr($filename, strlen($this->currentWorkingDirectory) + 1));
 		}
 
-		return $filename;
+		return str_replace('\\', '/', $filename);
```
Comment (staabm) [inline src/File/SimpleRelativePathHelper.php:21]: do we need the same fix in other implementations of `RelativePathHelper` ?

Comment (dktapps) [reply src/File/SimpleRelativePathHelper.php:21]: Not that I'm aware of. The one used for regular analysis already performs dirsep normalization.

**Comment by dktapps** [comment]
Test failure looks unrelated.

**Comment by ondrejmirtes** [comment]
Thank you!

**Pull request merged**